### ISSUE TITLE
Twitter bot

### ISSUE THREAD

**Issue by simondutertre** [ringcentral/engage-digital-chatbot-js#6]
Hello

Is is possible to use this bot library to do a twitter bot with dimelo ?

I am trying to do the same as your video but with twitter contents and i get 
```
status: 422,
statusText: 'Unprocessable Entity',
```

Thanks !

**Comment by zxdong262** [comment]
@simondutertre It do support twitter, check https://github.com/zxdong262/ping-pong-engage-digital-chatbot as example, if you send extra props or wrong props to content api, you may get this error.

**Comment by zxdong262** [comment]
@simondutertre Let me know if you still get errors, you may send more detail of your bot to my email: drake.zhao@ringcentral.com

**Comment by simondutertre** [comment]
I'm using your ping pong chatbot to test : 
I receive this tweet event : 

```
{ type: 'content.imported',
  id: '5d52b3f1bd0e545a76caba07',
  issued_at: '2019-08-13T12:58:25.982Z',
  resource:
   { type: 'twtr/tweet',
     id: '5d52b3f15256c1249fce234b',
     metadata:
      { tw_entities: [Object],
        approval_required: false,
        author_id: '5d514e2bbd0e54337a2591e7',
        body: '@foobarbaz yolo',
        body_input_format: 'text',
        category_ids: [],
        creator_id: null,
        date: '2019-08-13',
        first_in_thread: true,
        foreign_categories: [],
        foreign_id: '1161260561021247488',
        has_attachment: false,
        intervention_id: null,
        in_reply_to_author_id: null,
        in_reply_to_id: null,
        language: 'en',
        source_id: '5d4d646dd6cb00dee7a70f0a',
        status: 'new',
        thread_id: '5d52b3f15256c1249fce234c',
        thread_title: '@foobarbaz yolo',
        created_from: 'synchronizer',
        private: false } } }
```

i try to do 

```
  const reply = await client.reply(event, {
    in_reply_to_id: "5d4d646dd6cb00dee7a70f0a",
    body: 'this is a auto reply body',
  })
```
in you chatbot engine

and i get 
```{ message: 'No such content with id: "5d4d646dd6cb00dee7a70f0a"',
        error: 'not_found',
        status: 404 }```

**Comment by simondutertre** [comment]
I managed to answer to an email tho

**Comment by simondutertre** [comment]
I tried 
```
await client.reply(event, {
      body: 'this is a auto reply body',
      private: 1
    })
```
in index.js

and i got a 200 status with this body, but i don't have any response to my tweet :(
```
data:
   { id: '5d52da5b0e69dc612cbc910f',
     created_at: '2019-08-13T15:42:19Z',
     updated_at: '2019-08-13T15:42:19Z',
     approval_required: false,
     author_id: '5d4942cae2031161498cab5b',
     body: 'this is a auto reply body',
     body_input_format: 'text',
     category_ids: [],
     creator_id: '5d4943539200262efc523a7b',
     foreign_id: null,
     intervention_id: null,
     language: 'en',
     remotely_deleted: false,
     source_id: '5d4d646dd6cb00dee7a70f0a',
     source_url: null,
     synchronization_status: 'export_pending',
     status: 'user_reply',
     thread_id: '5d52da59eb702597d6ef8498',
     in_reply_to_id: '5d52da59eb702597d6ef8497',
     in_reply_to_author_id: '5d514e2bbd0e54337a2591e7',
     attachments: [],
     body_formatted:
      { text: 'this is a auto reply body',
        html: '<p>this is a auto reply body</p>' },
     created_from: 'api',
     private_message: true,
     published: true,
     source_type: 'twitter_search',
     type: 'private_tweet' }
```

**Comment by zxdong262** [comment]
I tested again, Dimelo chat/email works fine, do have problem with twitter source, only sync message when I save twitter source setting, but when synced, it works as expected. I send an email to exports@dimelo.com to ask about this, hope I can get some answer. And should not need set private: 1, inside there is a method determine whether it is private.

@simondutertre

**Comment by zxdong262** [comment]
@simondutertre You do follow this guide, right? https://github.com/ringcentral/engage-digital-chatbot-js/blob/master/docs/enable-twitter-source.md

**Comment by simondutertre** [comment]
Yes i followed this guide, i subscribed the webhook to content.imported on both an email source and a twitter source ( on a @ mention ) and i receive the tweets its just not replying

**Comment by zxdong262** [comment]
@simondutertre ping-pong bot only reply to "ping", you may test with private message only with "ping", or remove the isPing check.

**Comment by simondutertre** [comment]
Yes i removed the isPing or course 😁

Le mer. 14 août 2019 à 10:03, ZHAO Xudong <notifications@github.com> a
écrit :

**Comment by simondutertre** [comment]
If i understood you managee to respond to a tweet in private using your bot ? Could you give me a working example for it ?

**Comment by zxdong262** [comment]
<img width="1359" alt="Screen Shot 2019-08-14 at 4 08 49 PM" src="https://user-images.githubusercontent.com/1641949/63004837-de514b00-bead-11e9-93be-1fd71954912b.png">

**Comment by simondutertre** [comment]
Ok well the only difference is that my source on dimelo is a twitter search on a mention.

**Comment by zxdong262** [comment]
Mention may not work at all(quite a while ago, not very sure), you may try the message.

**Comment by simondutertre** [comment]
Ok i managed to do the same as you : 

![image](https://user-images.githubusercontent.com/21337263/63007285-a5e74800-be80-11e9-979a-d823f11b7afb.png)

Though my use case is not DM it is auto reply to mention, can you let me know if this will work ? 

Thanks a lot

**Comment by zxdong262** [comment]
@simondutertre I will test more about this.

**Comment by zxdong262** [comment]
@simondutertre Check this, https://github.com/zxdong262/ping-pong-engage-digital-chatbot/commit/a134bb84eae8b4fa6f792de9e90dfca003e33c30

**Comment by simondutertre** [comment]
Nice, it seems your code is doing the work but the message keeps staying in pending in dimelo interface, i haven't got my tweet response back yet : 

![image](https://user-images.githubusercontent.com/21337263/63012329-560d7e80-be8a-11e9-886d-f7022eef1962.png)

**Comment by simondutertre** [comment]
Did it respond to a mention on your part ?

**Comment by zxdong262** [comment]
@simondutertre Dimelo exports told me it may sync every 10 minutes due to rate limit, if you save source setting, it will trigger at once.

**Comment by simondutertre** [comment]
It is still in pending, no response to my tweet yet tho. @zxdong262

**Comment by zxdong262** [comment]
Oh I see, it is a twitter search, not a twitter, I may test it later, never used it before.

 @simondutertre

**Comment by simondutertre** [comment]
@zxdong262 Did you managed to test twitter search yet ?

**Comment by zxdong262** [comment]
@simondutertre Tested, but not sure how it works, asked dimelo exports about this, may test more next week.

**Comment by simondutertre** [comment]
@zxdong262 Ok, let me know when you have any info. Because the reason of my company dimelo subscription is integrating Twitter search unfortunately. Thanks you for your help

**Comment by simondutertre** [comment]
@zxdong262 Btw do you

**Comment by simondutertre** [comment]
@zxdong262 Oops ^^'

The chatbot framework doesnt integrate other action than reply right ? ( Setting a category / fetch a discussion thread )

**Comment by zxdong262** [comment]
@simondutertre Yes, other than reply, you could use client[VERB](url...), you may check https://github.com/ringcentral/engage-digital-client-js, and https://engage-api-docs.readthedocs.io/en/latest/

**Comment by simondutertre** [comment]
@zxdong262  any response from dimelo team ?

**Comment by simondutertre** [comment]
Got a response, they told me i need to add a body parameter with the @ i want to respond to like so :

https://csm1.api.engagement.dimelo.com/1.0/contents?access_token=1b685d40ac23ba0cbc244fd0&in_reply_to_id=5d51a332eb70253891d66a0e&body=@TelecomThibault

**Comment by simondutertre** [comment]
@zxdong262 
I just tested and their solution works

// POST request on : 

`https://csm1.api.engagement.dimelo.com/1.0/contents?access_token=xxxxxxxxx&in_reply_to_id=5d5ab762e203114699e9da32&body=@dutertresimon test123`

weird 'body' param tho

**Comment by zxdong262** [comment]
@simondutertre Thank you, I would test it!

**Comment by zxdong262** [comment]
@simondutertre I can not get the twitter search work, and Dimelo exports did not reply yet, can you give some guide about how to make search work and how to test it?

**Comment by simondutertre** [comment]
@zxdong262
 - You need access to the Standard Tiers Twitter Api for a twitter account. And then create an app ( https://developer.twitter.com/en/dashboard ). When asked about your usage put **Personal** and not **Entreprise** else you will be be asked questions about your usage and will need the Twitter team to validate them .. ( i waited 2 days for this )

- Once you have an app on https://developer.twitter.com/en/dashboard you can register the account as a new  'Twitter source'.

- Then you create a 'Twitter search source' and put the main account in 'Primary Twitter Source' 

![image](https://user-images.githubusercontent.com/21337263/63332405-d83ced80-c337-11e9-98ca-818436574721.png)

( notice i am testing the "AND" "API operators right now ^^ )

**Comment by zxdong262** [comment]
@simondutertre Thank you! I am applying the access, seems personal use still need to be validated.

**Comment by zxdong262** [comment]
@simondutertre I can get the twitter search source event now, but can not reply with API, Dimelo exports confirmed the API problem, are investigating the issue.

**Comment by simondutertre** [comment]
I tested it right now, it worked 
I did a POST on `https://csm1.api.engagement.dimelo.com/1.0/contents?access_token=xxxxxxxxx&in_reply_to_id=5d68eee5eb70253073eedb7d&body=@simonV003 test123`

**Comment by simondutertre** [comment]
Can you tell me what is are "handled" param and why you need isSelf function in your onEvent function ?

( is it to handle event dupplicates sent by Dimelo ? )

Thanks !

**Comment by zxdong262** [comment]
I find that in some situation webhook send twitter message twitted by self, in this case, we should not reply self.

`handled` is for bot skills, some skill may already handled the message and return true for handled, so user could choose ignore the message since it is already handled.

@simondutertre

**Comment by zxdong262** [comment]
If you are not using any bot skills, it is safe to ignore `handled`.
